### PR TITLE
WIP Routing strategy snippet 6.2

### PR DESCRIPTION
**Pull request by danielmarbach**
Relates to https://github.com/Particular/NServiceBus/pull/4461

@Particular/docs-maintainers wasn't exactly sure if we need more explanation. I went with the approach to have the same snippet versioned.

Since the old way still works and is supported would be still need an upgrade guide?

// cc @Particular/nservicebus-maintainers

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
I think it requires more explanation than just a single comment line. As a user I would not expect to see `NotImplementedException` in the working code.

**Comment by danielmarbach** [comment]
@SzymonPobiega so actually split the partial then? Fine for me

**Comment by SzymonPobiega** [comment]
@danielmarbach yeah. Maybe provide snippets for both cases in 6.2 partial: the simple one and the context-aware one.

**Comment by danielmarbach** [comment]
@SzymonPobiega pushed a change. Would that be what you had in mind?

File: nservicebus/messaging/routing-extensibility_distribution_core_[6,6.2).partial.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+include: routing-extensibility.distribution
```
Comment (SimonCropp) [inline nservicebus/messaging/routing-extensibility_distribution_core_[6,6.2).partial.md]: partials must match the version numbers of the snippets dirs. so unless u want to add a 6.2 here https://github.com/Particular/docs.particular.net/tree/master/Snippets/Core this partial will be ignored

**Comment by weralabaj** [comment]
@danielmarbach feel free to continue editing text, etc.
as discussed, I'll make sure we'll figure out the best way for moving forward from the engine/editing perspective

**Comment by danielmarbach** [comment]
@Particular/docs-maintainers @Particular/nservicebus-maintainers please review. I used a local nuget package to write this sample. The PR is merge into develop. We can always push an unstable package

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 2]
```diff
@@ -64,4 +64,6 @@ snippet:RoutingExtensibility-Distribution
 
 snippet:RoutingExtensibility-DistributionStrategy
 
+As of NServiceBus version 6.2 and later it is possible to override the virtual method `SelectDestination`. The method provides access to the `DistributionStrategyContext` that enables more advanced distribution scenarios like distributing based on the headers of the message. When `SelectDestination` is overridden, it is advised not to call `base.SelectDestination` since the base method calls into `SelectReceiver` for backward compatibility reasons. `SelectReceiver` can throw a `NotImplementedException` since it will never be called in this case.
```
Comment (SeanFeldman) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: > The method provides access to the `DistributionStrategyContext` that enables more advanced distribution scenarios like distributing based on the headers of the message.

replace "like" with "such as". Possibly:

The method provides access to the `DistributionStrategyContext` enabling more advanced distribution scenarios such as message header based distribution.

**Comment by weralabaj** [comment]
@danielmarbach 

> since the base method calls into `SelectReceiver` for backward compatibility reasons. `SelectReceiver` can throw a `NotImplementedException` since it will never be called in this case.

This part is a bit unclear to me. Is `SelectReceiver` part of the public API? Are we getting too deep into implementation details here? Is there actually any situation when you should call the base method? Or could we just say:

> When overridning `SelectDestination` do not call `base.SelectDestination`. For more details go to the sample.

**Comment by weralabaj** [comment]
see updated comment with propositions

**Comment by danielmarbach** [comment]
@weralabaj good suggestion! I think I dove into too much details, you are right. Updating based on your decision

File: nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md [hunk 3]
```diff
@@ -64,4 +64,6 @@ snippet:RoutingExtensibility-Distribution
 
 snippet:RoutingExtensibility-DistributionStrategy
 
-To learn more about creating custom distribution strategies see the [fair distribution sample](/samples/routing/fair-distribution/).
\ No newline at end of file
+In NServiceBus version 6.2 and above it is possible to override the virtual method `SelectDestination`. The method provides access to the `DistributionStrategyContext` that enables implementing more advanced distribution scenarios, such as distributing based on the headers of the message. When `SelectDestination` is overridden, do not call `base.SelectDestination` since the base method calls `SelectReceiver` for backward compatibility reasons. `SelectReceiver` can throw a `NotImplementedException`.
```
Comment (SimonCropp) [inline nservicebus/messaging/routing-extensibility_content_core_[6,).partial.md]: "NServiceBus" is assumed. should be removed

**Comment by weralabaj** [comment]
I think we need to deploy the 6.2 package to MyGet to get it working @danielmarbach

File: Snippets/Core/Core_6/Routing/Extensibility.cs [hunk 4]
```diff
@@ -181,4 +181,33 @@ class MyEvent :
         {
         }
     }
-}
\ No newline at end of file
+
+    public class Extensibility_62
+    {
+        #region RoutingExtensibility-DistributionStrategy [6.2,7.0)
```
Comment (SimonCropp) [inline Snippets/Core/Core_6/Routing/Extensibility.cs]: shouldnt this be an unbound upper?

File: Snippets/Core/Core_6/Routing/Extensibility.cs [hunk 5]
```diff
@@ -181,4 +181,33 @@ class MyEvent :
...
+
+        class RandomStrategy :
+            DistributionStrategy
+        {
+            static Random r = new Random();
```
Comment (SimonCropp) [inline Snippets/Core/Core_6/Routing/Extensibility.cs]: can we have a better variable name

**Comment by danielmarbach** [comment]
@weralabaj @SimonCropp addressed comments

**Pull request merged**